### ISSUE TITLE
Consider implementing from_offset/to_offset in UntaggedX classes for better initialization

### ISSUE THREAD

**Issue by sstrickl** [dart-lang/sdk#52876]
Currently, when the VM initializes instances of `Object` subclasses, it chooses one of the following options for initializing the allocated memory:

* `cid == kInstructionsCid`:  Initialize the memory by word with the target's breakpoint instruction (duplicated if smaller than word size).
* `IsTypedDataBaseClassId(cid)`: If the allocated object is from a new large page, do no initialization (as the page is zero-initialied by the OS), otherwise initialize the memory by word with 0.
* `cid == kArrayCid`: If the allocated object is from a new large page, do no initialization, otherwise initialize the memory by word with `Object::null()`.
* Otherwise, initialize the memory by word with `Object::null()`.

This ensures that the GC-visible fields of the object are initialized to GC-safe values (either `Object::null()` or a `SmiPtr` (0)).

However, not all fields of `Object` subclasses not listed above are `Ptr` fields. Initializing the object by word to `Object::null()` means those non-`Ptr` fields will be initialized to some portion of the address of `Object::null()`. If the `::New()` method for the class properly initializes all non-`Ptr` fields after calling `Object::Allocate()`, then this isn't an issue, but sometimes it is not, if the non-`Ptr` field is only used in certain circumstances. If it is not, and then serialized to snapshot, then now the snapshot contains a recorded field value that contains some portion of the address of `Object::null()` during compilation, which can result in non-deterministic builds.

Instead, we could consider implementing a system like the following, similar to how we handled objects with and without compressed fields semi-automatically:

* `VISIT_FROM(UntaggedX, field)` installs an extra static method `from_offset()`, which uses `OFFSET_OF(UntaggedX, field_)` to get the offset of the first `Ptr` field. (Unfortunately, as shown here, this requires passing the type of the class containing the static method to `VISIT_FROM`, since there's no way to get the type of the class in which the static member is defined as far as I can tell.)
* `VISIT_TO(UntaggedX, field)` installs an extra static method `to_offset()`, which does the same as `from_offset()` for the final `Ptr` field.
* Other `VISIT_FROM_X/VISIT_TO_X` versions are altered accordingly.
* `VISIT_NOTHING` installs versions of `from_offset()` and `to_offset()` which just return 0.
* Add static `from_offset()`/`to_offset()` methods on classes during `BASE_OBJECT_IMPLEMENTATION(X)` which call the `from_offset()`/`to_offset()` methods of the corresponding `UntaggedX` class.
* Add two new arguments to `Object::Allocate()` and `Object::InitializeObject` which are the values of calling `from_offset()` and `to_offset()` on the class (similar to how the last argument currently is the value of calling `ContainsCompressedPointers()` on the class).
* In `Object::InitializeObject`, initialize all memory after the object header up to the value of `from_offset()` with 0, the words between (and including) `from_offset()` and `to_offset()` to `Object::null()`, and all the words after `to_offset()` to the end of the allocated object to 0.

The special case for initializing `Instructions` objects and not initializing `TypedDataBase` and `Array` objects allocated from a new large page must be kept, but we could remove the special casing for the initial value to initialize memory to, except for one case: in `TypedDataBase` and its subclasses, the `length_` field is currently initialized to the `SmiPtr` 0, but without special casing it will instead be initialized to `Object::null()`, so either the `::New()` methods for its subclasses must ensure this field is initialized to 0 if that assumption is used by the code base, or if it's always set anyway then this doesn't really matter.

It's a bit of engineering, but this will remove the possibility in the future of non-`Ptr` fields contributing to non-determinism when snapshotting. 

/cc @rmacnak-google @mkustermann @mraleph for their thoughts and feedback.

**Comment by sstrickl** [comment]
A slight adjustment: the `from_offset()`/`to_offset()` methods from `VISIT_NOTHING()` need to return `target::kWordSize`, not `0`, because those methods need to return offsets that are _after_ the object header, which is not initialized as part of `InitializeObject`.

**Comment by sstrickl** [comment]
I haven't tried it yet, but now I'm curious whether you could do something like the following, similar to the definition of the `OFFSET_OF` macro:

```
template <T>
intptr_t untagged_from_offset() { 
  reinterpret_cast<intptr_t>(reinterpret_cast<T*>(kOffsetOfPtr)->from()) - kOffsetOfPtr;
}
template <T>
intptr_t untagged_to_offset() { 
  reinterpret_cast<intptr_t>(reinterpret_cast<T*>(kOffsetOfPtr)->to()) - kOffsetOfPtr;
}
```

that would avoid having to change all the `VISIT_FROM`/`VISIT_TO` definitions, except for `VISIT_NOTHING` which would need to define dummy versions of `from()` and `to()` that both return an address just past the object header, and then just add

```
#define BASE_OBJECT_IMPLEMENTATION(X) \
static intptr_t from_offset() { return untagged_from_offset<Untagged##X>(); } \
static intptr_t to_offset() { return untagged_to_offset<Untagged##X>(); } \
...
```

So the only places where we'd have to change each invocation would be the `Object::InitializeObject` calls, not the uses of `VISIT_FROM...`/`VISIT_TO...`.

**Comment by sstrickl** [comment]
Note we would need to have a similar `to_offset(intptr_t len)` for payload-containing objects, and could add a static constexpr field `kContainsPayload` to the untagged class, similar to `kContainsCompressedPointers`, that's set to `false` in `UntaggedObject` and inherited by subclasses unless `VISIT_TO_PAYLOAD_END()` is used, in which case an overriding definition set to `true` is created, to ensure that `to_offset()` is only called on classes whose objects don't contain payloads and `to_offset(intptr_t len)` is only called on classes of payload-containing objects.

**Comment by sstrickl** [comment]
After I realized it would require less changes than expected, I took an initial shot at this in [CL 312900](https://dart-review.googlesource.com/c/sdk/+/312900) to give an idea of what I expect the changes to look like. 

The main issue are the few times we allocate/initialize objects with only a cid: `Object::Clone` and two spots in `runtime/vm/object_graph_copy.cc`.

`Object::Clone` isn't actually an issue, since we'll copy the rest of the object over whatever Object::InitializeObject chooses to do, either initial value is GC safe until we finish the copying and return the clone, and no user of the cloned object will see any type-unsound contents until the copying is done. So not an issue.

While I'm not familiar with the `ObjectGraphCopier`, in this case I've just made sure we'd initialize objects created by it exactly the same way we did before: if the class is a subtype of TypedDataBase, then we'll initialize it like that (so only the length, which is set immediately afterwards, is set to `Object::null()` and the rest is set to 0), and otherwise we'll just initialize the entire contents of the object to `Object::null()` per the old behavior.
